### PR TITLE
feat: Adds docuploader tool

### PR DESCRIPTION
**Pull request by hemanshv**
The PR replaces the usage of python DocUploader to our in-house(.net) DocUploader.

### CODE REVIEW

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+using Google.Cloud.Tools.DocUploader;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+
+
+Parser.Default.ParseArguments<CreateMetadataOptions, UploadFileOptions>(args)
+    .WithParsed<CreateMetadataOptions>(options => CreateMetadata(options))
+    .WithParsed<UploadFileOptions>(options => UploadFile(options));
+
+// Creates the Metadata object based on the specified options and then store it as a JSON file at options.Destination.
```
Comment (anuragsrivstv) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:25]: This method does nothing to store data in `JSON` file.That decided by file format specified in `options.Destination` . may be we should update the comment accordingly.

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Uploads the zip file to the specified storage bucket.
+void UploadFile(UploadFileOptions options)
+{
+    string metadataFilePath = options.MetadataFile;
+    if (String.IsNullOrEmpty(metadataFilePath))
+    {
+        var jsonFile = Path.Combine(options.DocumentationPath, "docs.metadata.json");
+        if (File.Exists(jsonFile))
```
Comment (anuragsrivstv) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:51]: We can remove else block by somthing like :-


 if (!File.Exists(jsonFile))
        {
                        throw new Exception("You need metadata to upload the documents. You can generate it using the docuploader create-metadata command.");
        }
        metadataFilePath = jsonFile;

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            metadataFilePath = jsonFile;
+        }
+        else
+        {
+            throw new Exception("You need metadata to upload the docs.You can generate it with docuploader create-metadata");
```
Comment (anuragsrivstv) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:57]: May be instead of generic exception , it should be InvalidOperationException ?

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        }
+    }
+    if (Directory.EnumerateFiles(options.DocumentationPath).ToList().Count == 0)
+    {
+        throw new Exception($"The documentation path given({options.DocumentationPath}) does not contain any documentation files.");
```
Comment (anuragsrivstv) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:62]: May be we can re-write it as $"The documentation path `{options.DocumentationPath}` is empty. Please provide a path to a folder that contains documentation files."

File: tools/Google.Cloud.Tools.DocUploader/CreateMetadataOptions.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    [Option("xrefs", Required = false, Separator = ',')]
+    public IList<string> Xrefs { get; set; }
+
+    [Option("xref-services", Required = false, Separator = ',')]
+    public IList<string> XrefServices { get; set; }
+
+    [Option("destination", Default = "docs.metadata.json")]
+    public string Destination { get; set; }
+}
+
```
Comment (anuragsrivstv) [inline tools/Google.Cloud.Tools.DocUploader/CreateMetadataOptions.cs:57]: Extra lines.

File: tools/Google.Cloud.Tools.DocUploader/Metadata.g.cs [hunk 6]
```diff
@@ -0,0 +1,679 @@
+// <auto-generated>
+//     Generated by the protocol buffer compiler.  DO NOT EDIT!
+//     source: metadata.proto
+// </auto-generated>
+#pragma warning disable 1591, 0612, 3021, 8981
```
Comment (anuragsrivstv) [inline tools/Google.Cloud.Tools.DocUploader/Metadata.g.cs:5]: Just for my understanding, Is this file required ? if yes, why ?

File: tools/Google.Cloud.Tools.DocUploader/Upload.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+/// <summary>
+/// Helper class to upload a file to google cloud storage.
+/// </summary>
+public class Upload
+{
+    /// <summary>
+    /// Uploads the given file to the google cloud storage.
+    /// </summary>
+    /// <param name="source">The file to be uploaded on storage bucket. Must not be null.</param>
```
Comment (anuragsrivstv) [inline tools/Google.Cloud.Tools.DocUploader/Upload.cs:28]: As this is a documented, we should add null checks as well.

File: tools/Google.Cloud.Tools.DocUploader/Compress.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+using System.Diagnostics;
+
+namespace Google.Cloud.Tools.DocUploader;
+
+/// <summary>
+/// Helper class to perform the .tar.gz compression using powershell.
+/// </summary>
+public class Tar
+{
+    public static void Compress(string directory, string destination)
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Compress.cs:24]: Drive-by comment to start with: let's not run PowerShell and tar... it should be easy enough to create the tar file using SharpCompress.

File: tools/Google.Cloud.Tools.DocUploader/Upload.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    /// <param name="destination">The name of the object within the bucket. Must not be null.</param>
+    /// <param name="bucketName">The name of the bucket containing the object. Must not be null.</param>
+    /// <param name="credentialsPath">The path to credentials.json file. Must not be null.</param>
+    public static void UploadFile(string source, string destination, string bucketName, string credentialsPath)
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Upload.cs:32]: Given how little code there is here, I'm not sure it's actually worth having as a separate class... particularly if we end up going with creating the tarball in memory. I'd just inline this code, or at least move it to a local method in Program.

File: tools/Google.Cloud.Tools.DocUploader/UploadFileOptions.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    [Option("credentials", HelpText = "Path to the credentials file to use for Google Cloud Storage.")]
+    public string Credentials { get; set; }
+
+    [Option("metadata-file", HelpText = "Path to the docs.metadata file. The path must be relative to the CWD.")]
+    public string MetadataFile { get; set; }
+
+    [Option("destination-prefix", HelpText = "Prefix to include when uploading tar file. A - will be added after the prefix, if there is one.")]
+    public string DestinationPrefix { get; set; }
+}
+
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/UploadFileOptions.cs:36]: Nit: lots of blank lines

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using CommandLine;
+using Google.Cloud.Tools.DocUploader;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:19]: Nit: blank line

File: uploaddocs.sh [hunk 12]
```diff
@@ -47,10 +47,10 @@ do
 
     # TODO: Product page
     echo "Generating metadata (googleapis.dev)"
-    python -m docuploader create-metadata --name $pkg --version $version --language dotnet --github-repository googleapis/google-cloud-dotnet
+    dotnet run --project /tools/Google.Cloud.Tools.DocUploader create-metadata --name $pkg --version $version --language dotnet --github-repository googleapis/google-cloud-dotnet
```
Comment (jskeet) [inline uploaddocs.sh:50]: I'd suggest adding `--` after the arguments to dotnet, here and below. So:

```sh
dotnet run --project /tools/Google.Cloud.Tools.DocUploader -- create-metadata --name $pkg --version $version --language dotnet --github-repository googleapis/google-cloud-dotnet
```

It makes it clearer which arguments are expected to be parsed by `dotnet`, and which are expected to be parsed by our code.

**Review by jskeet — COMMENTED** [review]
Somewhat drive-by review for now (given the lex-review tag) - will wait for you to let me know you'd like me to take another pass.

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        throw new Exception($"The documentation path given({options.DocumentationPath}) does not contain any documentation files.");
+    }
+
+    var metadataText = File.ReadAllText(metadataFilePath);
+    var parser = new JsonParser(JsonParser.Settings.Default.WithIgnoreUnknownFields(true));
+    var metadata = parser.Parse<Metadata>(metadataText);
+
+    string tempFileName = "temp.tar.gz";
+    Tar.Compress(options.DocumentationPath, tempFileName);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.DocUploader/Program.cs:70]: Compute is our largest API, and that ends up at ~140MB uncompressed, and it compresses pretty well - down to under 10MB - so we should be fine to do this in memory.

File: tools/Google.Cloud.Tools.DocUploader/Program.cs (hunk 1 above)
Comment (hemanshv) [reply tools/Google.Cloud.Tools.DocUploader/Program.cs]: done

File: tools/Google.Cloud.Tools.DocUploader/Program.cs (hunk 2 above)
Comment (hemanshv) [reply tools/Google.Cloud.Tools.DocUploader/Program.cs]: done

File: tools/Google.Cloud.Tools.DocUploader/Compress.cs (hunk 8 above)
Comment (hemanshv) [reply tools/Google.Cloud.Tools.DocUploader/Compress.cs]: done

**Comment by hemanshv** [comment]
Hi @jskeet ,
I've incorporated the feedback from previous review. Please have a look and share your valuable feedback.

Thanks,

File: tools/Google.Cloud.Tools.DocUploader/CreateMetadataOptions.cs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+using Google.Protobuf.WellKnownTypes;
+
+[Verb("create-metadata", HelpText = "create metadata file")]
+internal class CreateMetadataOptions
+{
+    [Option("name", Required = true)]
+    public string PackageName { get; set; }
+
+    [Option("version", Required = true)]
+    public string version { get; set; }
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/CreateMetadataOptions.cs:25]: Nit: Version, not version (only noticed when reviewing Program.cs)

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        Name = options.PackageName,
+        Version = options.version,
+        Language = options.Language,
+        DistributionName = options.DistributionName,
+        ProductPage = options.ProductPage,
+        GithubRepository = options.GitHubRepo,
+        IssueTracker = options.IssueTracker,
+        Stem = options.Stem,
+    };
+    metadata.Xrefs.Add(options.Xrefs);
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:45]: These can be in the object initializer:

```csharp
var metadata = new Metadata
{
    // Code already above
    Xrefs = { options.Xrefs },
    XrefServices = { options.XrefServices }
}
```

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    metadata.XrefServices.Add(options.XrefServices);
+
+    File.WriteAllText(options.Destination, metadata.ToString());
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:48]: I may well create a PR to write out the indented JSON later on, but that definitely doesn't need to be done now.

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+// Uploads the zip file to the specified storage bucket.
+void CompressAndUploadFile(UploadFileOptions options)
+{
+    string tempFileName = "temp.tar.gz";
+    CompressDirectory(options.DocumentationPath, tempFileName);
+    UploadFile(tempFileName, GetFilenameFromMetadata(), options.StagingBucket, options.Credentials);
+
+    string GetFilenameFromMetadata()
+    {
+        string metadataFilePath = string.IsNullOrEmpty(options.MetadataFile) ? Path.Combine(options.DocumentationPath, "docs.metadata.json") : options.MetadataFile;
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:60]: It would be weird to specify an empty string - I think it's fine to use `options.MetadataFile ?? Path.Combine(options.DocumentationPath, "docs.metadata.json")`

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+/// <summary>
+/// Creates an instance of <see cref="Metadata"/> from json file.
+/// </summary>
+Metadata GetMetadataFromJsonFile(string metadataFilePath)
+{
+    if (!File.Exists(metadataFilePath))
+    {
+        throw new FileNotFoundException("You need metadata to upload the docs.You can generate it with docuploader create-metadata");
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:75]: Nit - space after the period. It might also be worth including which file we tried to read - here's an alternative:

```csharp
throw new UserErrorException($"Unable to load metadata file '{metadataFilePath}`. Metadata is required for upload; this can be generated using the create-metadata option.");

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+void CompressDirectory(string directory, string destination)
+{
+    if (Directory.EnumerateFiles(directory).Count() == 0)
+    {
+        throw new InvalidOperationException($"The documentation path `{directory}` is empty");
+    }
+
+    using var archive = TarArchive.Create();
+    archive.AddAllFromDirectory(directory);
+    var twopt = new TarWriterOptions(CompressionType.GZip, finalizeArchiveOnClose: true);
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:107]: Let's rename this to just `writerOptions`

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    archive.SaveTo(destination, twopt);
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:108]: This is where we could use a MemoryStream if we wanted to do it all in memory.

File: tools/Google.Cloud.Tools.DocUploader/Program.cs [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    metadata.Xrefs.Add(options.Xrefs);
+    metadata.XrefServices.Add(options.XrefServices);
+
+    File.WriteAllText(options.Destination, metadata.ToString());
+}
+
+// Uploads the zip file to the specified storage bucket.
+void CompressAndUploadFile(UploadFileOptions options)
+{
+    string tempFileName = "temp.tar.gz";
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.DocUploader/Program.cs:54]: Please use Path.GetTempFileName (and then possibly write it out on the console) if you definitely want to use a physical file.

File: uploaddocs.sh [hunk 22]
```diff
@@ -47,10 +47,10 @@ do
 
     # TODO: Product page
     echo "Generating metadata (googleapis.dev)"
-    python -m docuploader create-metadata --name $pkg --version $version --language dotnet --github-repository googleapis/google-cloud-dotnet
+    dotnet run --project /tools/Google.Cloud.Tools.DocUploader -- create-metadata --name $pkg --version $version --language dotnet --github-repository googleapis/google-cloud-dotnet
```
Comment (jskeet) [inline uploaddocs.sh:50]: I think it would be better to give a repo-relative path here. Using /tools isn't ideal, IMO.

You could use `REPOROOT=$(git rev-parse --show-toplevel)` to obtain the repository root directory at the start of this script. (We might want to figure out if we could install this as a dotnet tool, but that's a different matter.)

File: docs/generate-devsite-help.sh [hunk 23]
```diff
@@ -36,24 +36,24 @@ cd output/devsite-help
 
 # Create the docs metadata. We assume we may need to refer to any of the utility libraries.
 # We also refer to Datastore, PubSub and Storage.
-python -m docuploader create-metadata \
+dotnet run --project ../../../tools/Google.Cloud.Tools.DocUploader -- create-metadata \
```
Comment (jskeet) [inline docs/generate-devsite-help.sh:39]: You need these changes in generate-devsite-utilities.sh too.

(And as elsewhere, I think it would be good to use the repo root as an anchor, although at least this is relative.)

**Review by jskeet — COMMENTED** [review]
Code is almost entirely fine now - just a few tweaks here and there, and in the shell scripts.

File: tools/Google.Cloud.Tools.DocUploader/Program.cs (hunk 17 above)
Comment (hemanshv) [reply tools/Google.Cloud.Tools.DocUploader/Program.cs]: done

File: tools/Google.Cloud.Tools.DocUploader/Program.cs (hunk 21 above)
Comment (hemanshv) [reply tools/Google.Cloud.Tools.DocUploader/Program.cs]: Changed the compression logic to hold the compressed file in-memory.

File: docs/generate-devsite-help.sh (hunk 23 above)
Comment (hemanshv) [reply docs/generate-devsite-help.sh]: done

Comment (jskeet) [reply docs/generate-devsite-help.sh]: I'm not seeing that file in the PR... missed a "git add" perhaps?

Comment (hemanshv) [reply docs/generate-devsite-help.sh]: oh I think I missed that, let me do that.

Comment (hemanshv) [reply docs/generate-devsite-help.sh]: updated it please have a look.

**Comment by jskeet** [comment]
I'll look again on Tuesday (I'm about to go on vacation) - or @amanda-tarafa could do so earlier than that.

**Review by jskeet — APPROVED** [review]
Looks good, thanks! I'll start investigating how hard it would be to publish this as a dotnet tool...
